lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
   return nullptr;
 }
 
-/// Removes common tail enable operations for sequential 'then'/'else'
-/// branches inside an 'if' operation.
-///
-///   if %a with %A {                       if %a with %A {
-///     seq { ... calyx.enable @B }           seq { ... }
-///   else {                          ->    } else {
-///     seq { ... calyx.enable @B }           seq { ... }
-///   }                                     }
-///                                         calyx.enable @B
-static LogicalResult eliminateCommonTailEnable(IfOp ifOp,
-                                               PatternRewriter &rewriter) {
-  // Check if the branches exist.
+/// Returns a mapping of {Enabled Group, EnableOp} for all EnableOps within the
+/// immediate ParOp's body.
+static llvm::StringMap<EnableOp> getAllEnableOpsInImmediateBody(ParOp parent) {
+  llvm::StringMap<EnableOp> enables;
+  Block *body = parent.getBody();
+  for (EnableOp op : body->getOps<EnableOp>())
+    enables.insert(std::pair(op.groupName(), op));
+
+  return enables;
+}
+
+/// Removes common tail enable operations for then and else regions encompassed
+/// in SeqOp and ParOp. This canonicalization is stringent about not entering
+/// nested control operations, as this may cause unintentional changes in
+/// behavior.
+static LogicalResult eliminateCommonTail(IfOp ifOp, PatternRewriter &rewriter) {
   if (!ifOp.thenRegionExists() || !ifOp.elseRegionExists())
     return failure();
+  rewriter.setInsertionPointAfter(ifOp);
 
-  auto &thenOpStructureOp = ifOp.getThenBody()->front();
-  auto &elseOpStructureOp = ifOp.getElseBody()->front();
-  // TODO(circt/#1861): ParOps have less restrictive conditions.
-  if (isa<ParOp>(thenOpStructureOp) || isa<ParOp>(elseOpStructureOp))
-    return failure();
-
-  // At this point, only sequential operations are valid inside the branches.
-  auto thenSeqOp = dyn_cast<SeqOp>(thenOpStructureOp);
-  auto elseSeqOp = dyn_cast<SeqOp>(elseOpStructureOp);
-  assert(thenSeqOp && elseSeqOp &&
-         "expected nested seq ops in both branches of a calyx.IfOp");
-
-  EnableOp lastThenEnableOp = getLastEnableOp(thenSeqOp);
-  EnableOp lastElseEnableOp = getLastEnableOp(elseSeqOp);
+  auto &thenControl = ifOp.getThenBody()->front();
+  auto &elseControl = ifOp.getElseBody()->front();
+  if (auto r1 = dyn_cast<ParOp>(thenControl);
```

Comment:
In this case we essentially have 2 separate canonicalization patterns with disjoint pattern match conditions. Again, it might be personal preference, but i would prefer having two distinct rewrite patterns to capture each of the match (and rewrite) conditions.

Reply:
I agree. My rationale for using the simple `hasCanonicalizeMethod` was it is easier when there's just one match and rewrite. But now that we are adding a separate way of matching and rewriting, I think using `hasCanonicalizer` with two separate patterns is preferable. If that makes sense, perhaps a first step would be a NFC to convert from `hasCanonicalizeMethod`->`hasCanonicalizer`, and then this PR would be rebased onto that and add a second pattern.